### ISSUE TITLE
When updating, download & verify new package before uninstalling

### ISSUE THREAD

**Issue by spongbros** [lukesampson/scoop#2220]
Currently, performing an upgrade of a package runs an uninstallation before downloading the updated files. It is possible for the download to fail or for the checksum to not match, leaving no copy of the package installed and, potentially, no way to reinstall it. Ideally, the update package should be downloaded and verified before the uninstallation process begins.

**Comment by quincunx** [comment]
The current behaviour is unfortunate but an old version is not remove at all.

E.g. I've vim installed - which gets an update every 1-2 days. Every single installed/updated version is still on my system, see the directory ``%USERPROFILE%\scoop\apps\vim``:

```
8.0.1765
8.0.1766
8.0.1772
8.0.1774
8.0.1780
8.0.1787
8.0.1789
8.0.1795
current
```

``current`` is a link to the latest installed version (8.0.1795).

So all it would need is to detect a failed update and act accordingly. Maybe uninstalling (of shims) and removing the ``current`` link is not required at all - as it will be overwritten by the new version (if it's successful).

https://github.com/lukesampson/scoop/wiki/The-'Current'-Version-Alias

**Comment by r15ch13** [comment]
When a download or hash check fails it will leave an empty directory behind. If we remove it and simply run `scoop reset <app>` it should be fixed.

**Comment by gitolicious** [comment]
Maybe `reset` could check if `current` links to an empty folder and fix it automatically.
Also, are `post_install` steps performed and are all shims put back in place with the `reset` command? Otherwise the app will end up in a messed up state.

**Comment by r15ch13** [comment]
@gitolicious no, reset doesn't run `post_intall`, it's only for fixing shortcuts, shims, and environment variables. See [scoop-reset.ps1](https://github.com/lukesampson/scoop/blob/master/libexec/scoop-reset.ps1)

**Comment by gitolicious** [comment]
Well, then unfortunately your suggestion is only valid for very basic manifests. I see two solutions here:
1. Perform download to cache before uninstallation -> should prevent most internet connection related problems
2. Perform self-healing by re-installing the old version after a failed update -> brings the user back to a usable state
Probably both can solutions can be put in place in parallel.

**Comment by gitolicious** [comment]
PR for solution 1 in #2343

**Comment by r15ch13** [comment]
`reset` is mostly for switching between package versions like `python` and `python27`, which have the same executables. Running `post_install` is not suited for this command.

**Comment by pawelkuras** [comment]
Just experienced this in https://github.com/lukesampson/scoop-extras/issues/1269 but it is not the first time.  Considering how many issues come up when you search for **is:issue "hash check fail"** on the main scoop repo, not to mention the buckets, this would be very nice to have.

**Comment by pawelkuras** [comment]
Coming back to the number of issues raised.  I think a lot less "hash check fail" issues would be raised if the app wasn't left in an uninstalled state, saving time fixing hashes.
